### DISCUSSION TITLE
Support URI base path for hosting behind reverse proxies (e.g. Home Assistant)

### DISCUSSION THREAD

**Discussion by matt-FFFFFF** [teslamate-org/teslamate#494] [Ideas]
**Is your feature request related to a problem? Please describe.**

TeslaMate does not support the use of a base path element and expects all requests to be relative to the root ```/```. This causes problems when hosting behind a reverse proxy in a shared environment. My use case is integration with Home Assistant and wishing to use the ingress capability. The benefits of this are being able to leverage the existing TLS and AuthN capabilities of the HA platform.

**Describe the solution you'd like**

I would like to be able to specify a BASE_PATH system environment variable that is read at runtime. The TeslaMate endpoint would then respond to requests relative to this base path.

e.g.
```
docker run --env BASE_PATH=/my/custom/path teslamate/teslamate:1.17
```

The alternative with Home Assistant is to map the port onto the network and access TeslaMate directly, however this means there is no AuthN nor TLS. This makes porting the docker container to a HA addon much more complex as we need to implement these features independently.

I have tried to see if I could implement this myself and do a PR, but I cannot yet get my head around Elixir!

**Additional context**

[Home Assistant Ingress](https://developers.home-assistant.io/docs/hassio_addon_presentation#ingress)

**Comment by cowlinb6** [comment]
Has anyone found a solution to using subpaths yet?

**Comment by matt-FFFFFF** [comment]
Without support in the code there isn't a workaround I don't think. TM generates internal links that are based on the site root (/).

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by adriankumpf** [comment]
As a workaround, does your setup allow to run e.g. nginx in front of TeslaMate to rewrite the paths?

**Comment by matt-FFFFFF** [comment]
I did try this, I think some resources paths are generated by TM and the paths are all relative to the site root.

E.g. the CSS and the internal links within the admin interface

**Comment by matt-FFFFFF** [comment]
I have tried to explain the issue here @adriankumpf 

The ingress correctly rewrites the requests inbound but does not rewrite the references in the response. I could investigate using [sub_filter](http://nginx.org/en/docs/http/ngx_http_sub_module.html) but this seems messy.

![diagram](https://www.websequencediagrams.com/cgi-bin/cdraw?lz=dGl0bGUgSG9tZSBBc3Npc3Rhbm50IEluZ3Jlc3MKCkJyb3dzZXItPkhBAA0IOiBHRVQgL2FkZG9ucy88AAMFaWQ-CgAYCi0-VGVzbGFNYXRlACYHCgAICQA8DkhUTUwgY29udGVudAA0DQBxBwATDwCBBQkAFglFdmFsdWF0ZSByZXNvdXJjZXMgKGUuZy4gQ1NTL2pzKQCBJhtjc3MvZmlsZS5jc3MAZhZFcnJvciA0MDQKCg&s=default)

**Comment by tcm0116** [comment]
I spent some time yesterday playing around with sub-filter to see if I can get this working. My setup was very basic as I was only really able to add a listener to the existing HA NGINX add-on that I'm using for SSL access. 

Despite my simple setup, I was able to get the main screen working via the NGINX proxy. It's a little slow to load the first time because sub-filter requires that compression be disabled in the response sent to the client, but subsequent requests are fast due to cashing.

However, I'm having a bit of an issue with the Geo-Fence and Settings pages as they continually refresh, keeping the mouse in a loading cursor, which prevents clicking on anything. I think this might be complicated by the fact that I'm using a http to https proxy, so my next step is to try and get a dedicated proxy setup for this.

@matt-FFFFFF - when you were experimenting with this, how were you running NGINX? It seems to me that the cleanest way would be to incorporate it into the TeslaMate docker image since it would be part of the TeslaMate service as a whole, but that's not really the "Docker way". I did notice that you're already using an S6 base image, so maybe it does make sense to go ahead and incorporate NGINX into the TeslaMate docker image. Thoughts?

**Comment by tcm0116** [comment]
@matt-FFFFFF and I were able to get Ingress mostly working by using a combination of NGINX sub-filtering and some modifications to the runtime configuration of TeslaMate. The details can be see at: https://github.com/matt-FFFFFF/hassio-addon-teslamate/pull/1 and https://github.com/matt-FFFFFF/hassio-addon-teslamate/pull/12

The one thing that doesn't seem to be working is the breadcrumb navigation at the top of the Settings and Geo-Fence pages. Those links always want to point back to the base URL excluding any path that has been configured. I believe those are controlled by a Phoenix live view, so the TeslaMate code would have to be modified to respect the new root path. I may start looking into this, but I wanted to document the things that would need to be done or looked into to enable a configurable root path for TeslaMate without having to use sub-filtering.

- Modify runtime.exs to pass a path parameter to the url configuration for TeslaMateWeb.Endpoint. I imagine this path would be configurable via an environment variable, which would default to "/" if the environment variable is not set. With this parameter, many of the URLs generated by TeslaMate will include the specified path.
- Look into why the TESLAMATE link in the black title bar does not utilize the path specified in the url configuration.
- Look into why the font paths in the the CSS do not utilize the path specified in the url configuration.
- Look into why the websocket path and the "/js/" path for geo-fence in the javascriupt file do not utilize the path specified in the url configuration.
- Look into why the breadcrumb navigation in the Settings and Geo-Fence pages do not utilize the path specified in the url configuration.

**Reply by matt-FFFFFF** [reply]
Thanks again to @tcm0116 for this work. I honestly did very little and Thomas deserves the credit here!

**Comment by ileodo** [comment]
I managed to reverse proxy teslamate behind a sub url by using `sub_filter`

here is the segment in my nginx config. (assuming host it under sub path `/teslamate/`

```
    sub_filter 'href="/' 'href="/teslamate/';
    sub_filter 'src="/' 'src="/teslamate/';

    sub_filter 'url(/' 'url(/teslamate/';
    sub_filter '"/live"' '"/teslamate/live"';
    sub_filter '"/js/"' '"/teslamate/js/"';
    sub_filter 'baseUrl:window.location.origin' 'baseUrl:window.location.origin+"/teslamate"';
    sub_filter 'url:this.redirect?void 0:this.href||void 0' 'url:this.redirect?void 0:(this.href!==null?this.href.replace("/teslamate","/"):this.href)||void 0';
    sub_filter 'onRedirect({to:o,flash:r}){this.liveSocket.redirect(o,r)}' 'onRedirect({to:o,flash:r}){this.liveSocket.redirect(o.replace(o.startsWith("/")?/^\//gm:/\/${dollar_sign}/gm, "/teslamate/"),r)}';
    
```

you also need 

```
geo $dollar_sign {
        default "$";
}
```

**Reply by ileodo** [reply]
basically I believe the last 3 sub_filter already indicate the places in the source code should be changed to support sub_url properly.
